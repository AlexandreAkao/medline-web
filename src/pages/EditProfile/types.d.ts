type EditUserRegisterTypes = 'name' | 'password' | 'email';

interface IEditUserRegisterAction {
  type: EditUserRegisterTypes;
  payload: string;
}
