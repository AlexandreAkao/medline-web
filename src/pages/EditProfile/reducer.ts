export const editUserFormReducer = (state: IEditUserRegister, action: IEditUserRegisterAction) => {
  switch (action.type) {
    case 'name':
    case 'password':
    case 'email':
      return { ...state, [action.type]: action.payload };
    default:
      return state;
  }
};

export const editUserFormInitialState = {
  name: '',
  password: '',
  email: '',
};
