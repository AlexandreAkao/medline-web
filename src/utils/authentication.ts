import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
