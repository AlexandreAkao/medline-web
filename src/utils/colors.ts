export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
