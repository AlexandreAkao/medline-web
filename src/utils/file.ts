export const downloadFile = (blob: Blob, filename: string) => {
  const link = document.createElement('a');
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  link.click();
};
