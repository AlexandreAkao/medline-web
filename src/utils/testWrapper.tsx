import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component: React.ReactNode) =>
  render(<MemoryRouter>{component}</MemoryRouter>);
