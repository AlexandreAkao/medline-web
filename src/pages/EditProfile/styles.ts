import styled from 'styled-components';

import colors from 'styles/colors';
import { Container } from 'components/TextInput/styles';
import { ButtonContainer } from 'components/Button/styles';

export const EditProfileContainer = styled.main`
  background-color: ${colors.primary.normal};
  height: 100%;
  overflow: auto;
`;

export const EditProfileSection = styled.section`
  margin: 5rem auto 0 auto;
  width: 50%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  ${Container} {
    margin: 1rem 0;
  }

  ${ButtonContainer} {
    margin-top: 2rem;
  }
`;

export const EditProfileTitle = styled.h1`
  font-size: 2rem;
  color: ${colors.white.normal};
`;
