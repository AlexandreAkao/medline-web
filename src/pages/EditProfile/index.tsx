import { ChangeEvent, useEffect, useReducer, useState } from 'react';
import { toast } from 'react-toastify';
import { useNavigate } from 'react-router-dom';

import { EditProfileContainer, EditProfileSection, EditProfileTitle } from 'pages/EditProfile/styles';
import MedlineHeader from 'components/Header/MedlineHeader';
import TextInput from 'components/TextInput';
import Button from 'components/Button';
import { editUserFormInitialState, editUserFormReducer } from 'pages/EditProfile/reducer';
import api from 'service/api';
import { useAuth } from 'hooks/useAuth';

function EditProfile() {
  const [confirmPassword, setConfirmPassword] = useState('');
  const [editUserForm, dispatchEditUserForm] = useReducer(editUserFormReducer, editUserFormInitialState);
  const navigate = useNavigate();
  const { saveUserInfo } = useAuth();

  useEffect(() => {
    api.get<IUser>('user/info').then(data => {
      const { name, email } = data.data;

      dispatchEditUserForm({ payload: name, type: 'name' });
      dispatchEditUserForm({ payload: email, type: 'email' });
    });
  }, []);

  const handleChangeInput = (type: EditUserRegisterTypes) => (event: ChangeEvent<HTMLInputElement>) =>
    dispatchEditUserForm({ payload: event.target.value, type });

  const handleSubmitEditUser = async () => {
    if (editUserForm.password !== confirmPassword || editUserForm.password === '' || confirmPassword === '') {
      toast.error('As senhas precisam ser iguais');
      return;
    }
    const data = await api.put('user', editUserForm);

    if (data.status === 200) {
      toast.success('Usu√°rio atualizado com sucesso!');
      saveUserInfo(data.data);
      navigate('/');
    }
  };

  return (
    <EditProfileContainer>
      <MedlineHeader />

      <EditProfileSection>
        <EditProfileTitle>Editar Perfil</EditProfileTitle>
        <TextInput label="Nome" placeholder="Nome" value={editUserForm.name} onChange={handleChangeInput('name')} />
        <TextInput label="Email" placeholder="Email" value={editUserForm.email} onChange={handleChangeInput('email')} />

        <TextInput label="Senha" placeholder="Senha" onChange={handleChangeInput('password')} isPassword />
        <TextInput
          label="Confirmar senha"
          placeholder="Confirmar senha"
          onChange={e => setConfirmPassword(e.target.value)}
          isPassword
        />
        <Button primary={false} size="large" onClick={handleSubmitEditUser}>
          Salvar
        </Button>
      </EditProfileSection>
    </EditProfileContainer>
  );
}

export default EditProfile;
