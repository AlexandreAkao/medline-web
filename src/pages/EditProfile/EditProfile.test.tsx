import { screen } from '@testing-library/react';

import { renderWithProvider } from 'utils/testWrapper';
import EditProfile from 'pages/EditProfile';

describe('EditProfile', () => {
  const renderComponent = () => {
    return renderWithProvider(<EditProfile />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    renderComponent();
    const titleElement = screen.getByRole('heading');
    const inputElements = screen.getAllByRole('textbox');
    const buttonElements = screen.getAllByRole('button');
    expect(titleElement).toHaveTextContent('Editar Perfil');
    expect(inputElements).toHaveLength(4);
    expect(buttonElements).toHaveLength(2);
  });
});
