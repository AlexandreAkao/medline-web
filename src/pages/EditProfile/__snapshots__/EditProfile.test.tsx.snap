// Vitest Snapshot v1

exports[`EditProfile > should match with snapshot 1`] = `
<div>
  <main
    class="sc-jSMfEi sDxaw"
  >
    <header
      class="sc-papXJ bicVkd"
      data-testid="test-navigation"
    >
      <img
        alt="Medline logo"
        class="sc-jqUVSM dEwOGF"
        src="/src/assets/images/medline-logo.svg"
      />
      <nav
        class="sc-kDDrLX ekWlzz"
        data-testid="test-navigation-items"
      >
        <div
          class="sc-ftvSup lpdRfb"
        >
          <button
            class="sc-eCYdqJ fDdwce"
          >
            Cadastro
          </button>
        </div>
        <div
          class="sc-ftvSup lpdRfb"
        >
          <a
            href="/login"
          >
            Entrar
          </a>
        </div>
      </nav>
    </header>
    <section
      class="sc-gKXOVf kQmpnx"
    >
      <h1
        class="sc-iBkjds eYqXUC"
      >
        Editar Perfil
      </h1>
      <div
        class="sc-bczRLJ itlDIl"
      >
        <span
          class="sc-gsnTZi jYHEuj"
          data-testid="test-input-label"
        >
          Nome
        </span>
        <div
          class="sc-dkzDqf bMJUCZ"
        >
          <input
            class="sc-hKMtZM fIniom"
            data-testid="test-input"
            placeholder="Nome"
            role="textbox"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="sc-bczRLJ itlDIl"
      >
        <span
          class="sc-gsnTZi jYHEuj"
          data-testid="test-input-label"
        >
          Email
        </span>
        <div
          class="sc-dkzDqf bMJUCZ"
        >
          <input
            class="sc-hKMtZM fIniom"
            data-testid="test-input"
            placeholder="Email"
            role="textbox"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="sc-bczRLJ itlDIl"
      >
        <span
          class="sc-gsnTZi jYHEuj"
          data-testid="test-input-label"
        >
          Senha
        </span>
        <div
          class="sc-dkzDqf bMJUCZ"
        >
          <input
            class="sc-hKMtZM fIniom"
            data-testid="test-input"
            placeholder="Senha"
            role="textbox"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="sc-bczRLJ itlDIl"
      >
        <span
          class="sc-gsnTZi jYHEuj"
          data-testid="test-input-label"
        >
          Confirmar senha
        </span>
        <div
          class="sc-dkzDqf bMJUCZ"
        >
          <input
            class="sc-hKMtZM fIniom"
            data-testid="test-input"
            placeholder="Confirmar senha"
            role="textbox"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="sc-eCYdqJ kYzDoT"
      >
        Salvar
      </button>
    </section>
  </main>
  ,
</div>
`;
